ravedude/src/board.rs

```diff
@@ -1,439 +1,89 @@
-        _ => return None,
-    })
-}
-
-// ----------------------------------------------------------------------------
-
-fn find_port_from_vid_pid_list(list: &[(u16, u16)]) -> anyhow::Result<std::path::PathBuf> {
-    for serialport::SerialPortInfo {
-        port_name,
-        port_type,
-    } in serialport::available_ports().unwrap()
-    {
-        if let serialport::SerialPortType::UsbPort(usb_info) = port_type {
-            for (vid, pid) in list.iter() {
-                if usb_info.vid == *vid && usb_info.pid == *pid {
-                    return Ok(port_name.into());
-                }
-            }
-        }
-    }
-    Err(anyhow::anyhow!("Serial port not found."))
-}
-
-// ----------------------------------------------------------------------------
-
-struct ArduinoUno;
-
-impl Board for ArduinoUno {
-    fn display_name(&self) -> &str {
-        "Arduino Uno"
-    }
-
-    fn needs_reset(&self) -> Option<&str> {
-        None
-    }
-
-    fn avrdude_options(&self) -> avrdude::AvrdudeOptions {
-        avrdude::AvrdudeOptions {
-            programmer: "arduino",
-            partno: "atmega328p",
-            baudrate: None,
-            do_chip_erase: true,
-        }
-    }
-
-    fn guess_port(&self) -> Option<anyhow::Result<std::path::PathBuf>> {
-        Some(find_port_from_vid_pid_list(&[
-            (0x2341, 0x0043),
-            (0x2341, 0x0001),
-            (0x2A03, 0x0043),
-            (0x2341, 0x0243),
-        ]))
-    }
-}
-
-struct ArduinoMicro;
-
-impl Board for ArduinoMicro {
-    fn display_name(&self) -> &str {
-        "Arduino Micro"
-    }
-
-    fn needs_reset(&self) -> Option<&str> {
-        Some("Reset the board by pressing the reset button once.")
-    }
-
-    fn avrdude_options(&self) -> avrdude::AvrdudeOptions {
-        avrdude::AvrdudeOptions {
-            programmer: "avr109",
-            partno: "atmega32u4",
-            baudrate: Some(115200),
-            do_chip_erase: true,
-        }
-    }
-
-    fn guess_port(&self) -> Option<anyhow::Result<std::path::PathBuf>> {
-        Some(find_port_from_vid_pid_list(&[
-            (0x2341, 0x0037),
-            (0x2341, 0x8037),
-            (0x2A03, 0x0037),
-            (0x2A03, 0x8037),
-            (0x2341, 0x0237),
-            (0x2341, 0x8237),
-        ]))
-    }
-}
-
-struct ArduinoNano;
-
-impl Board for ArduinoNano {
-    fn display_name(&self) -> &str {
-        "Arduino Nano"
-    }
-
-    fn needs_reset(&self) -> Option<&str> {
-        None
-    }
-
-    fn avrdude_options(&self) -> avrdude::AvrdudeOptions {
-        avrdude::AvrdudeOptions {
-            programmer: "arduino",
-            partno: "atmega328p",
-            baudrate: Some(57600),
-            do_chip_erase: true,
+use std::collections::HashMap;
+
+use crate::config;
+
+fn get_all_boards() -> anyhow::Result<HashMap<String, config::BoardConfig>> {
+    toml::from_str(include_str!("boards.toml")).map_err(|err| {
```

Comment:
To me this `include_str!()` is fine, I wouldn't waste time trying to parse the file at compile-time.

We should just make sure your testcase actually gets run in CI to catch errors in contributed pull-requests early.  Can you adjust `.github/workflows/ci.yml` to also run `cargo test` for ravedude, in addition to `cargo check` please?

Reply:
Done :+1: